File: stackunderflow/etl/normalize/cursor.py
"""Cursor IDE normalizer.

Cursor v3 stores chat bubbles in a SQLite vscdb with **no per-message
token counts** — every ``tokenCount.{inputTokens,outputTokens}`` is
zero. The existing adapter (``stackunderflow/adapters/cursor.py``)
falls back to ``len(text) // 4`` for those rows and stamps
``cost_source = 'estimated'`` on the raw payload. The normalizer
mirrors that policy:

* Prefer explicit token counts when **either** field is non-zero —
  stamp ``cost_source='rate_card'``.
* Otherwise estimate ``len(content_text) // 4`` for the input side
  (assistant rows still need a price point, even if we under-count) —
  stamp ``cost_source='estimated'``.
* Model: prefer the value already lifted onto the row by the adapter
  (``parsed.modelInfo.modelName`` or ``providerOptions.cursor.modelName``).
  Fall back to the configured Cursor default ``'composer-1'`` when the
  row carries the adapter's own ``'cursor-auto'`` placeholder.

Provider-specific provenance worth preserving in ``raw_extras``: the
``composerData`` reference, ``conversationId``, and the
``cost_source`` flag the adapter already stamped (so a downstream
reader can distinguish adapter-level estimation from normalizer-level
estimation, even though the two should always agree).
"""

from __future__ import annotations

import json
from collections.abc import Iterable

from stackunderflow.infra.costs import RATE_CARD

from .base import (
    COST_SOURCE_ESTIMATED,
    COST_SOURCE_RATE_CARD,
    COST_SOURCE_UNKNOWN,
    Normalizer,
)

# Default model when the adapter wrote the placeholder ``"cursor-auto"`` —
# Cursor's modern default agent model. Spec: PR description.
_DEFAULT_MODEL = "composer-1"

# Keys we copy out of the persisted raw payload into ``raw_extras`` so
# downstream consumers can still trace back to Cursor's bubble.
_RAW_EXTRAS_FIELDS = ("conversationId", "composerData", "cost_source")


class CursorNormalizer(Normalizer):
    provider_name = "cursor"

    def normalize(self, msg_row: dict) -> Iterable[dict]:
        role = str(msg_row.get("role") or "")
        if role != "assistant":
            return

        input_tokens, output_tokens, estimated = _resolve_tokens(msg_row)

        # If neither real nor estimated tokens land us with a non-zero
        # number — and the row also has no text to estimate from — drop
        # it. A pure-empty assistant message is not billable.
        if input_tokens == 0 and output_tokens == 0:
            return

        model = _resolve_model(msg_row)

        if estimated:
            cost_source = COST_SOURCE_ESTIMATED
        elif model in RATE_CARD:
            cost_source = COST_SOURCE_RATE_CARD
        else:
            cost_source = COST_SOURCE_UNKNOWN

        raw_extras = _extras_from_raw_json(msg_row.get("raw_json"))

        yield self._build_event(
            msg_row,
            input_tokens=input_tokens,
            output_tokens=output_tokens,
            cache_read_tokens=int(msg_row.get("cache_read_tokens") or 0),
            cache_create_tokens=int(msg_row.get("cache_create_tokens") or 0),
            cost_source=cost_source,
            model=model,
            raw_extras=raw_extras,
        )


# ── helpers ─────────────────────────────────────────────────────────


def _resolve_tokens(msg_row: dict) -> tuple[int, int, bool]:
    """Return ``(input, output, estimated)``.

    Decision order:
      1. ``tokenCount.inputTokens`` / ``tokenCount.outputTokens`` from
         either the msg_row directly (synthetic test fixtures) or the
         persisted raw payload — when at least one is non-zero, these
         are authoritative and ``estimated=False``.
      2. Pre-canonicalised columns ``input_tokens`` / ``output_tokens``
         when at least one is non-zero — also authoritative.
      3. Estimate ``input = len(content_text) // 4``, ``output = 0``.
         Cursor v3 doesn't differentiate prompt vs. completion text on
         a single bubble, so estimation goes on the input side only —
         that mirrors the existing adapter logic and matches the test
         contract.
    """
    explicit = _explicit_token_count(msg_row)
    if explicit is not None:
        inp, out = explicit
        return inp, out, False

    inp_col = int(msg_row.get("input_tokens") or 0)
    out_col = int(msg_row.get("output_tokens") or 0)
    if inp_col > 0 or out_col > 0:
        return inp_col, out_col, False

    text = str(msg_row.get("content_text") or "")
    return max(len(text) // 4, 0), 0, True


def _explicit_token_count(msg_row: dict) -> tuple[int, int] | None:
    """Return ``(input, output)`` from a ``tokenCount`` block if set.

    Synthetic dicts can pass ``tokenCount`` directly on msg_row. Real
    rows surface the same block as ``raw_json.tokenCount``.
    """
    tc = msg_row.get("tokenCount")
    if not isinstance(tc, dict):
        payload = _safe_load_raw(msg_row.get("raw_json"))
        if isinstance(payload, dict):
            tc = payload.get("tokenCount")
    if not isinstance(tc, dict):
        return None
    inp = int(tc.get("inputTokens", 0) or 0)
    out = int(tc.get("outputTokens", 0) or 0)
    if inp == 0 and out == 0:
        return None
    return max(inp, 0), max(out, 0)


def _resolve_model(msg_row: dict) -> str:
    """Pick the most specific model id available, defaulting to
    ``composer-1`` when only the adapter's ``cursor-auto`` placeholder
    is present.
    """
    direct = msg_row.get("model")
    if isinstance(direct, str) and direct and direct != "cursor-auto":
        return direct

    payload = _safe_load_raw(msg_row.get("raw_json"))
    if isinstance(payload, dict):
        info = payload.get("modelInfo")
        if isinstance(info, dict):
            name = info.get("modelName")
            if isinstance(name, str) and name:
                return name
        opts = payload.get("providerOptions")
        if isinstance(opts, dict):
            cursor_opts = opts.get("cursor")
            if isinstance(cursor_opts, dict):
                name = cursor_opts.get("modelName")
                if isinstance(name, str) and name:
                    return name

    return _DEFAULT_MODEL


def _extras_from_raw_json(raw_json: object) -> dict | None:
    payload = _safe_load_raw(raw_json)
    if not isinstance(payload, dict):
        return None
    out: dict = {}
    for key in _RAW_EXTRAS_FIELDS:
        val = payload.get(key)
        if val is not None and val != "":
            out[key] = val
    return out or None


def _safe_load_raw(raw_json: object) -> object | None:
    if isinstance(raw_json, dict):
        return raw_json
    if not isinstance(raw_json, str | bytes | bytearray):
        return None
    try:
        if isinstance(raw_json, bytes | bytearray):
            return json.loads(raw_json.decode("utf-8", errors="replace"))
        return json.loads(raw_json)
    except (json.JSONDecodeError, ValueError, UnicodeDecodeError):
        return None

File: rust/crates/stax-etl/src/normalize/cursor.rs
//! Cursor IDE — port of `etl/normalize/cursor.py`.
//!
//! Cursor v3 stores chat bubbles in a vscdb with **no** per-message token
//! counts — every `tokenCount.{inputTokens,outputTokens}` is zero — so the
//! adapter falls back to `len(text) // 4` and stamps `estimated` on the raw
//! payload. This mirrors that policy on the normalizer side, and preserves the
//! adapter's own flag in `raw_extras` so a reader can tell adapter-level
//! estimation from normalizer-level estimation even though the two should
//! always agree.
//!
//! The model resolution ladder ends at `composer-1` — Cursor's modern default —
//! whenever the row still carries the adapter's `cursor-auto` placeholder.

use super::base::{CostSource, EventSpec, NormalizeContext, Normalizer, UsageEvent};
use super::row::{
    MsgRow, PyRaise, as_dict, as_nonempty_str, int_or_zero, int_or_zero_value, safe_load_raw,
    str_or_empty,
};
use super::text::{estimate_from_text, extras_from_payload};

/// The default when the adapter wrote the `cursor-auto` placeholder.
const DEFAULT_MODEL: &str = "composer-1";
const RAW_EXTRAS_FIELDS: [&str; 3] = ["conversationId", "composerData", "cost_source"];

/// The `cursor` normalizer.
#[derive(Debug, Clone, Copy, Default)]
pub struct CursorNormalizer;

impl Normalizer for CursorNormalizer {
    fn provider_name(&self) -> &'static str {
        "cursor"
    }

    fn normalize(&self, ctx: &NormalizeContext, row: &MsgRow) -> Result<Vec<UsageEvent>, PyRaise> {
        if str_or_empty(row, "role") != "assistant" {
            return Ok(Vec::new());
        }

        let (input_tokens, output_tokens, estimated) = resolve_tokens(row)?;

        // Neither real nor estimated tokens, and no text to estimate from: a
        // pure-empty assistant message is not billable.
        if input_tokens == 0 && output_tokens == 0 {
            return Ok(Vec::new());
        }

        let model = resolve_model(row);
        let cost_source = if estimated {
            CostSource::Estimated
        } else {
            ctx.rate_card_or_unknown(&model)
        };

        Ok(vec![
            self.build_event(
                ctx,
                row,
                EventSpec::new(
                    input_tokens,
                    output_tokens,
                    // Cursor's cache columns pass straight through — unlike
                    // copilot's, which are hard-zeroed.
                    int_or_zero(row, "cache_read_tokens")?,
                    int_or_zero(row, "cache_create_tokens")?,
                    cost_source,
                )
                .model(model)
                .raw_extras(extras_from_payload(row.get("raw_json"), &RAW_EXTRAS_FIELDS)),
            ),
        ])
    }
}

/// `(input, output, estimated)`.
///
/// An explicit `tokenCount` block wins, then the canonical columns when either
/// is positive, then `len(content_text) // 4` on the **input** side only —
/// Cursor v3 does not separate prompt from completion text on a single bubble.
fn resolve_tokens(row: &MsgRow) -> Result<(i64, i64, bool), PyRaise> {
    if let Some((input, output)) = explicit_token_count(row)? {
        return Ok((input, output, false));
    }
    let input_column = int_or_zero(row, "input_tokens")?;
    let output_column = int_or_zero(row, "output_tokens")?;
    if input_column > 0 || output_column > 0 {
        return Ok((input_column, output_column, false));
    }
    let text = str_or_empty(row, "content_text");
    Ok((estimate_from_text(&text), 0, true))
}

/// `(input, output)` from a `tokenCount` block, or `None` when both are zero.
fn explicit_token_count(row: &MsgRow) -> Result<Option<(i64, i64)>, PyRaise> {
    let payload = safe_load_raw(row.get("raw_json"));
    let block = match as_dict(row.get("tokenCount")) {
        Some(direct) => Some(direct.clone()),
        None => payload
            .as_ref()
            .and_then(|p| as_dict(p.get("tokenCount")))
            .cloned(),
    };
    let Some(block) = block else {
        return Ok(None);
    };
    // `int(tc.get("inputTokens", 0) or 0)` — the default and the `or` are
    // redundant with each other; both are ported.
    let input = int_or_zero_value(block.get("inputTokens"))?;
    let output = int_or_zero_value(block.get("outputTokens"))?;
    if input == 0 && output == 0 {
        return Ok(None);
    }
    Ok(Some((input.max(0), output.max(0))))
}

/// The most specific model id available.
fn resolve_model(row: &MsgRow) -> String {
    if let Some(direct) = as_nonempty_str(row.get("model"))
        && direct != "cursor-auto"
    {
        return direct.to_string();
    }
    if let Some(payload) = safe_load_raw(row.get("raw_json")) {
        if let Some(info) = as_dict(payload.get("modelInfo"))
            && let Some(name) = as_nonempty_str(info.get("modelName"))
        {
            return name.to_string();
        }
        if let Some(options) = as_dict(payload.get("providerOptions"))
            && let Some(cursor) = as_dict(options.get("cursor"))
            && let Some(name) = as_nonempty_str(cursor.get("modelName"))
        {
            return name.to_string();
        }
    }
    DEFAULT_MODEL.to_string()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::normalize::test_support::{assistant_row, ctx};
    use stax_core::queries::pyjson::Value as PyValue;

    fn cursor_row() -> MsgRow {
        assistant_row("cursor", "cursor-auto")
    }

    #[test]
    fn an_explicit_token_count_block_wins_and_is_not_estimated() {
        let row = cursor_row().with(
            "raw_json",
            PyValue::Str(r#"{"tokenCount": {"inputTokens": 90, "outputTokens": 10}}"#.into()),
        );
        let events = CursorNormalizer.normalize(&ctx(), &row).unwrap();
        assert_eq!((events[0].input_tokens, events[0].output_tokens), (90, 10));
        assert_ne!(events[0].cost_source, CostSource::Estimated);
    }

    #[test]
    fn an_all_zero_token_count_block_falls_through_to_estimation() {
        let row = cursor_row()
            .with(
                "raw_json",
                PyValue::Str(r#"{"tokenCount": {"inputTokens": 0, "outputTokens": 0}}"#.into()),
            )
            .with("content_text", PyValue::Str("x".repeat(80)));
        let events = CursorNormalizer.normalize(&ctx(), &row).unwrap();
        assert_eq!((events[0].input_tokens, events[0].output_tokens), (20, 0));
        assert_eq!(events[0].cost_source, CostSource::Estimated);
    }

    #[test]
    fn the_placeholder_model_resolves_to_composer_1() {
        let row = cursor_row().with("content_text", PyValue::Str("x".repeat(40)));
        let events = CursorNormalizer.normalize(&ctx(), &row).unwrap();
        assert_eq!(events[0].model, "composer-1");
    }

    #[test]
    fn the_model_ladder_reads_modelinfo_then_provideroptions() {
        let base = cursor_row().with("content_text", PyValue::Str("x".repeat(40)));
        let info = base.clone().with(
            "raw_json",
            PyValue::Str(r#"{"modelInfo": {"modelName": "claude-opus-4-7"}}"#.into()),
        );
        assert_eq!(
            CursorNormalizer.normalize(&ctx(), &info).unwrap()[0].model,
            "claude-opus-4-7"
        );
        let options = base.with(
            "raw_json",
            PyValue::Str(r#"{"providerOptions": {"cursor": {"modelName": "gpt-5"}}}"#.into()),
        );
        assert_eq!(
            CursorNormalizer.normalize(&ctx(), &options).unwrap()[0].model,
            "gpt-5"
        );
    }

    #[test]
    fn an_empty_bubble_is_not_billable() {
        assert!(
            CursorNormalizer
                .normalize(&ctx(), &cursor_row())
                .unwrap()
                .is_empty()
        );
    }

    #[test]
    fn the_adapters_own_cost_source_flag_is_preserved_verbatim() {
        let row = cursor_row().with("input_tokens", PyValue::Int(10)).with(
            "raw_json",
            PyValue::Str(r#"{"conversationId": "c-1", "cost_source": "estimated"}"#.into()),
        );
        let events = CursorNormalizer.normalize(&ctx(), &row).unwrap();
        assert_eq!(
            events[0].raw_extras.as_deref(),
            Some(r#"{"conversationId": "c-1", "cost_source": "estimated"}"#)
        );
    }
}
